pkg/scalers/datadog_scaler.go

```diff
@@ -0,0 +1,316 @@
+	appKey      string
+	datadogSite string
+	query       string
+	queryValue  int
+	vType       valueType
+	metricName  string
+	age         int
+}
+
+var datadogLog = logf.Log.WithName("datadog_scaler")
+
+// NewDatadogScaler creates a new Datadog scaler
+func NewDatadogScaler(ctx context.Context, config *ScalerConfig) (Scaler, error) {
+	meta, err := parseDatadogMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Datadog metadata: %s", err)
+	}
+
+	apiClient, err := newDatadogConnection(ctx, meta)
+	if err != nil {
+		return nil, fmt.Errorf("error establishing Datadog connection: %s", err)
+	}
+	return &datadogScaler{
+		metadata:  meta,
+		apiClient: apiClient,
+	}, nil
+}
+
+func parseDatadogMetadata(config *ScalerConfig) (*datadogMetadata, error) {
+	meta := datadogMetadata{}
+
+	if val, ok := config.TriggerMetadata["query"]; ok {
+		meta.query = val
+	} else {
+		return nil, fmt.Errorf("no query given")
+	}
+
+	if val, ok := config.TriggerMetadata["queryValue"]; ok {
+		queryValue, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("queryValue parsing error %s", err.Error())
+		}
+		meta.queryValue = queryValue
+	} else {
+		return nil, fmt.Errorf("no queryValue given")
+	}
+
+	if val, ok := config.TriggerMetadata["age"]; ok {
+		age, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("age parsing error %s", err.Error())
+		}
+		meta.age = age
+	} else {
+		meta.age = 90 // Default window 90 seconds
+	}
+
+	// For all the points in a given window, we take the rollup to the window size
+	rollup := fmt.Sprintf(".rollup(avg, %d)", meta.age)
+	meta.query = meta.query + rollup
+
+	if val, ok := config.TriggerMetadata["type"]; ok {
+		val = strings.ToLower(val)
+		switch val {
+		case "average":
+			meta.vType = average
+		case "global":
+			meta.vType = global
+		default:
+			return nil, fmt.Errorf("type has to be global or average")
+		}
+	} else {
+		meta.vType = average // Default to average between pods
+	}
+
+	if val, ok := config.TriggerMetadata["metricName"]; ok {
+		meta.metricName = GenerateMetricNameWithIndex(config.ScalerIndex, kedautil.NormalizeString(fmt.Sprintf("datadog-%s", val)))
+	} else {
+		return nil, fmt.Errorf("no metric name given")
+	}
+
+	if val, ok := config.AuthParams["apiKey"]; ok {
+		meta.apiKey = val
+	} else {
+		return nil, fmt.Errorf("no api key given")
+	}
+
+	if val, ok := config.AuthParams["appKey"]; ok {
+		meta.appKey = val
+	} else {
+		return nil, fmt.Errorf("no app key given")
+	}
+
+	if val, ok := config.AuthParams["datadogSite"]; ok {
+		if val != "" {
+			meta.datadogSite = val
+		} else {
+			meta.datadogSite = "datadoghq.com"
+		}
+	} else {
+		meta.datadogSite = "datadoghq.com"
+	}
+
+	return &meta, nil
+}
+
+// newDatddogConnection tests a connection to the Datadog API
+func newDatadogConnection(ctx context.Context, meta *datadogMetadata) (*datadog.APIClient, error) {
+
+	ctx = context.WithValue(
+		context.Background(),
```

Comment:
why not use the given `ctx` parameter here rather than creating a new context root?

```suggestion
		ctx,
```

Reply:
+1